File: gateway.py
import socket
import threading
import re
import base64
from http import HTTPStatus

SESSION_COOKIE_NAME = "auth_token"

PORT = 2222
TOKEN = "letmein"
SOCKET_DIR = "/tmp"
SOCKET_SUFFIX = ".sock"
DOMAIN = ".wkube.iiasa.ac.at"

def extract_subdomain(host_header: bytes) -> str:
    try:
        host = host_header.decode().split(":")[0]  # remove port
        if host.endswith(DOMAIN):
            subdomain = host.replace(DOMAIN, "")
            return subdomain
    except Exception as e:
        print(f"Failed to parse subdomain: {e}")
    return None

def connect_unix_socket(subdomain):
    path = f"{SOCKET_DIR}/{subdomain}{SOCKET_SUFFIX}"
    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    s.connect(path)
    return s


def parse_headers(data):
    headers = {}
    lines = data.split(b"\r\n")
    for line in lines[1:]:
        if b":" in line:
            k, v = line.split(b":", 1)
            headers[k.strip().lower()] = v.strip()
    return headers

def extract_subdomain(host: bytes) -> str:
    host = host.decode().split(":")[0]
    if host.endswith(DOMAIN):
        return host.replace(DOMAIN, "")
    return None

def is_valid_cookie(headers):
    cookies = headers.get(b'cookie')
    if not cookies:
        return False
    cookie_str = cookies.decode()
    for part in cookie_str.split(";"):
        if part.strip().startswith(f"{SESSION_COOKIE_NAME}="):
            val = part.strip().split("=")[1]
            if val == TOKEN:
                return True
    return False

def is_valid_basic_auth(headers):
    auth = headers.get(b'authorization')
    if not auth or not auth.startswith(b"Basic "):
        return False
    encoded = auth.split(b" ", 1)[1]
    try:
        decoded = base64.b64decode(encoded).decode()
        username, password = decoded.split(":", 1)
        return password == TOKEN
    except Exception:
        return False

def handle_http(client):
    data = client.recv(4096)
    if not data:
        client.close()
        return

    headers = parse_headers(data)
    host_header = headers.get(b"host")
    if not host_header:
        client.send(b"HTTP/1.1 400 Bad Request\r\n\r\nMissing Host header")
        client.close()
        return

    subdomain = extract_subdomain(host_header)
    if not subdomain:
        client.send(b"HTTP/1.1 400 Bad Request\r\n\r\nInvalid Host")
        client.close()
        return

    # Check for cookie or basic auth
    if not is_valid_cookie(headers) and not is_valid_basic_auth(headers):
        client.send(
            b"HTTP/1.1 401 Unauthorized\r\n"
            b"WWW-Authenticate: Basic realm=\"Access Required\"\r\n"
            b"Content-Type: text/plain\r\n\r\n"
            b"Unauthorized: Valid token required\n"
        )
        client.close()
        return

    # If basic auth was valid but no cookie, issue a Set-Cookie header
    set_cookie = b""
    if is_valid_basic_auth(headers) and not is_valid_cookie(headers):
        set_cookie = (
            f"Set-Cookie: {SESSION_COOKIE_NAME}={TOKEN}; Path=/; HttpOnly\r\n".encode()
        )

    # Forward to backend
    try:
        backend = connect_unix_socket(subdomain)
        backend.sendall(data)

        # Read response from backend
        response_data = b""
        while True:
            chunk = backend.recv(4096)
            if not chunk:
                break
            response_data += chunk

        backend.close()

        # Inject Set-Cookie header into response (optional)
        if set_cookie:
            parts = response_data.split(b"\r\n\r\n", 1)
            if len(parts) == 2:
                headers, body = parts
                headers += b"\r\n" + set_cookie
                response_data = headers + b"\r\n\r\n" + body

        client.sendall(response_data)
    except Exception as e:
        client.send(b"HTTP/1.1 502 Bad Gateway\r\n\r\nBackend error\n")
        print(f"[HTTP] Proxy error: {e}")
    finally:
        client.close()
def handle_tcp(client):
    try:
        client.send(b"Token: ")
        token = client.recv(1024).strip()
        if token.decode() != TOKEN:
            client.send(b"Access denied\n")
            client.close()
            return

        client.send(b"Access granted. Connecting...\n")

        # Get SNI-style domain (hacky): read peeked Host header manually
        peek = client.recv(256, socket.MSG_PEEK)
        match = re.search(rb'Host: ([a-zA-Z0-9\-]+)\.wkube\.iiasa\.ac\.at', peek)
        if not match:
            client.send(b"No valid Host found for backend\n")
            client.close()
            return

        subdomain = match.group(1).decode()

        backend = connect_unix_socket(subdomain)

        def forward(src, dst):
            try:
                while True:
                    data = src.recv(4096)
                    if not data:
                        break
                    dst.sendall(data)
            finally:
                src.close()
                dst.close()

        threading.Thread(target=forward, args=(client, backend)).start()
        threading.Thread(target=forward, args=(backend, client)).start()

    except Exception as e:
        print(f"[TCP] Error: {e}")
        client.close()

def handle_client(client):
    try:
        peek = client.recv(16, socket.MSG_PEEK)
        if peek.startswith(b"GET") or peek.startswith(b"POST") or b"HTTP" in peek:
            handle_http(client)
        else:
            handle_tcp(client)
    except Exception as e:
        print(f"Connection error: {e}")
        client.close()

def main():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(("0.0.0.0", PORT))
    s.listen(50)
    print(f"Gateway listening on port {PORT}...")

    while True:
        client, _ = s.accept()
        threading.Thread(target=handle_client, args=(client,)).start()

if __name__ == "__main__":
    main()

File: gateway.go
// Full Go conversion of the Python gateway with complete socket forwarding logic (hardened)
package main

import (
	"bufio"
	"bytes"
	"crypto/tls"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"
)

const (
	Port              = ":6473"
	SessionCookieName = "auth_token"
	SocketDir         = "/tmp"
	SocketSuffix      = ".sock"
	DomainSuffix      = ".wkube.iiasa.ac.at"
	JWKSUrl           = "https://accelerator-api.iiasa.ac.at/.well-known/jwks.json"
	ExpectedKid       = "rsa-key-2024-07"
	CertPath          = "/etc/ssl/certs/tls.crt"
	KeyPath           = "/etc/ssl/certs/tls.key"
)

var (
	// Reused TLS config, loaded once.
	tlsConfig *tls.Config

	// Precompiled regex for subdomain sanity checks (used by SSH).
	subdomainRe = regexp.MustCompile(`^[a-zA-Z0-9\-]+$`)

	// A list of common HTTP methods for quick plaintext detection.
	httpMethods = [][]byte{
		[]byte("GET "), []byte("POST "), []byte("PUT "),
		[]byte("DELETE "), []byte("HEAD "), []byte("OPTIONS "),
		[]byte("PATCH "), []byte("TRACE "), []byte("CONNECT "),
	}
)

// bufferedConn lets us Peek without losing bytes, then present a net.Conn whose Read drains from the bufio.Reader first.
type bufferedConn struct {
	net.Conn
	br *bufio.Reader
}

func (b *bufferedConn) Read(p []byte) (int, error) { return b.br.Read(p) }

func initTLSConfig() error {
	cert, err := tls.LoadX509KeyPair(CertPath, KeyPath)
	if err != nil {
		return fmt.Errorf("load keypair: %w", err)
	}
	tlsConfig = &tls.Config{
		Certificates: []tls.Certificate{cert},
		MinVersion:   tls.VersionTLS12,
		// Keep a sensible, modern baseline; let Go negotiate ECDHE suites.
		// Explicit list optional; these are fine for RSA certs:
		CipherSuites: []uint16{
			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
		},
	}
	return nil
}

func extractSubdomain(host string) string {
	host = strings.Split(host, ":")[0]
	if strings.HasSuffix(host, DomainSuffix) {
		return strings.TrimSuffix(host, DomainSuffix)
	}
	return ""
}

func connectUnixSocket(subdomain string) (net.Conn, error) {
	path := filepath.Join(SocketDir, subdomain+SocketSuffix)
	return net.Dial("unix", path)
}

// bidirectionalCopy wires src<->dst until both directions finish.
// It does not close the passed conns itself; the caller owns closing.
func bidirectionalCopy(a, b net.Conn) {
	var wg sync.WaitGroup
	wg.Add(2)

	// Direction a -> b
	go func() {
		defer wg.Done()
		_, _ = io.Copy(b, a)
		// Try half-close if supported to signal EOF without tearing down both ways.
		type closeWriter interface{ CloseWrite() error }
		if cw, ok := b.(closeWriter); ok {
			_ = cw.CloseWrite()
		}
	}()

	// Direction b -> a
	go func() {
		defer wg.Done()
		_, _ = io.Copy(a, b)
		if cw, ok := a.(closeWriter); ok {
			_ = cw.CloseWrite()
		}
	}()

	wg.Wait()
}

// handleHTTP terminates TLS (if any) before calling this, and then proxies HTTP to the backend Unix socket.
func handleHTTP(conn net.Conn) {
	defer conn.Close()

	// Parse the HTTP request robustly.
	br := bufio.NewReader(conn)
	req, err := http.ReadRequest(br)
	if err != nil {
		log.Printf("[HTTP] read request error: %v", err)
		_, _ = io.WriteString(conn, "HTTP/1.1 400 Bad Request\r\nContent-Length: 0\r\n\r\n")
		return
	}

	host := req.Host
	if host == "" {
		log.Printf("[HTTP] missing Host header")
		_, _ = io.WriteString(conn, "HTTP/1.1 400 Bad Request\r\n\r\nMissing Host header")
		return
	}

	subdomain := extractSubdomain(host)
	if subdomain == "" {
		log.Printf("[HTTP] invalid host %q", host)
		_, _ = io.WriteString(conn, "HTTP/1.1 400 Bad Request\r\n\r\nInvalid Host")
		return
	}

	// --- Auth/JWT checks intentionally omitted (explicitly removed) ---

	backend, err := connectUnixSocket(subdomain)
	if err != nil {
		log.Printf("[HTTP] backend connect failed: subdomain=%s err=%v", subdomain, err)
		_, _ = io.WriteString(conn, "HTTP/1.1 502 Bad Gateway\r\n\r\nBackend connection failed")
		return
	}
	defer backend.Close()

	// Ensure the Host header is preserved.
	req.URL.Scheme = ""        // proxying raw HTTP/1.1
	req.URL.Host = ""          // must be empty for RequestURI form
	req.RequestURI = req.URL.RequestURI()

	// Forward the parsed request to the backend.
	if err := req.Write(backend); err != nil {
		log.Printf("[HTTP] write to backend failed: subdomain=%s err=%v", subdomain, err)
		return
	}

	// After initial request/headers/body, tunnel both ways (for upgrades like WebSocket).
	bidirectionalCopy(conn, backend)
}

// handleTLS wraps the connection with TLS using the global tlsConfig, then delegates to handleHTTP.
func handleTLS(conn net.Conn) {
	if tlsConfig == nil {
		log.Println("[TLS] tlsConfig is nil (not initialized)")
		_ = conn.Close()
		return
	}
	tlsConn := tls.Server(conn, tlsConfig)
	if err := tlsConn.Handshake(); err != nil {
		log.Printf("[TLS] handshake error: %v", err)
		_ = tlsConn.Close()
		return
	}
	handleHTTP(tlsConn)
}

// handlePlainHTTP reads a plaintext HTTP request and 301-redirects to HTTPS while preserving the path/query.
func handlePlainHTTP(conn net.Conn) {
	defer conn.Close()

	br := bufio.NewReader(conn)
	req, err := http.ReadRequest(br)
	if err != nil {
		log.Printf("[HTTP-PLAIN] read request error: %v", err)
		_, _ = io.WriteString(conn, "HTTP/1.1 400 Bad Request\r\nContent-Length: 0\r\n\r\n")
		return
	}

	if req.Host == "" {
		_, _ = io.WriteString(conn, "HTTP/1.1 400 Bad Request\r\n\r\nMissing Host header")
		return
	}

	location := "https://" + req.Host + req.URL.RequestURI()
	resp := fmt.Sprintf("HTTP/1.1 301 Moved Permanently\r\nLocation: %s\r\nContent-Length: 0\r\nConnection: close\r\n\r\n", location)
	if _, err := io.WriteString(conn, resp); err != nil {
		log.Printf("[HTTP-PLAIN] failed to write redirect: %v", err)
	}
}

// handleSSH provides a tiny prompt, validates subdomain, and then wires stdin/stdout to the Unix socket.
func handleSSH(conn net.Conn) {
	defer conn.Close()

	reader := bufio.NewReader(conn)
	_, _ = conn.Write([]byte("Subdomain: "))
	subdomain, err := reader.ReadString('\n')
	if err != nil {
		_, _ = conn.Write([]byte("Read error\n"))
		return
	}
	subdomain = strings.TrimSpace(subdomain)
	if !subdomainRe.MatchString(subdomain) {
		_, _ = conn.Write([]byte("Invalid subdomain format\n"))
		return
	}

	_, _ = conn.Write([]byte("Connecting...\n"))
	backend, err := connectUnixSocket(subdomain)
	if err != nil {
		log.Printf("[SSH] backend connect failed: subdomain=%s err=%v", subdomain, err)
		_, _ = conn.Write([]byte("Backend connection failed\n"))
		return
	}
	defer backend.Close()

	// Wire up both directions until EOF.
	bidirectionalCopy(conn, backend)
}

func looksLikeHTTP(peek []byte) bool {
	for _, m := range httpMethods {
		if bytes.HasPrefix(peek, m) {
			return true
		}
	}
	// Fallback heuristic.
	return bytes.Contains(peek, []byte("HTTP/"))
}

func handleConnection(conn net.Conn) {
	// Short classification timeout to avoid slowloris on first read.
	_ = conn.SetReadDeadline(time.Now().Add(2 * time.Second))
	br := bufio.NewReader(conn)
	peek, err := br.Peek(8)
	if err != nil {
		_ = conn.Close()
		return
	}
	_ = conn.SetReadDeadline(time.Time{}) // clear deadline

	// Wrap so subsequent reads drain from br (including peeked bytes).
	bconn := &bufferedConn{Conn: conn, br: br}

	switch {
	case looksLikeHTTP(peek):
		log.Printf("[CONN] plaintext HTTP from %s", conn.RemoteAddr())
		handlePlainHTTP(bconn)
		return

	// TLS record header: 0x16 (Handshake), 0x03 (TLS major), next is minor {0x01..0x04}
	case len(peek) >= 3 && peek[0] == 0x16 && peek[1] == 0x03:
		log.Printf("[CONN] TLS from %s", conn.RemoteAddr())
		handleTLS(bconn)
		return

	case bytes.HasPrefix(peek, []byte("SSH-")):
		log.Printf("[CONN] SSH from %s", conn.RemoteAddr())
		handleSSH(bconn)
		return

	default:
		_, _ = bconn.Write([]byte("Unrecognized protocol. Only SSH and TLS are supported.\n"))
		_ = bconn.Close()
	}
}

func main() {
	log.SetFlags(log.LstdFlags | log.Lmicroseconds)
	if err := initTLSConfig(); err != nil {
		log.Fatalf("[BOOT] TLS init failed: %v", err)
	}

	ln, err := net.Listen("tcp", Port)
	if err != nil {
		log.Fatalf("[BOOT] listen error: %v", err)
	}
	log.Println("[BOOT] Gateway listening on", Port)

	for {
		conn, err := ln.Accept()
		if err != nil {
			log.Printf("[ACCEPT] error: %v", err)
			continue
		}
		go handleConnection(conn)
	}
}
